server/src/main/java/io/druid/client/selector/ServerSelector.java

```diff
@@ -62,59 +66,88 @@ public void addServerAndUpdateSegment(
   {
     synchronized (this) {
       this.segment.set(segment);
-      int priority = server.getServer().getPriority();
-      Set<QueryableDruidServer> priorityServers = servers.computeIfAbsent(priority, p -> new HashSet<>());
+      Set<QueryableDruidServer> priorityServers = new HashSet<>();
+      if (server.getServer().getType() == ServerType.REALTIME) {
+        priorityServers = realtimeServers.computeIfAbsent(
+            server.getServer().getPriority(), p -> new HashSet<>());
+      } else if (server.getServer().getType() == ServerType.HISTORICAL) {
+        priorityServers = historicalServers.computeIfAbsent(
+            server.getServer().getPriority(), p -> new HashSet<>());
+      }
```

Comment:
Please add else clause with throwing exception, to not silently consume potential problem